pilot/pkg/proxy/envoy/v2/lds.go

```diff
@@ -65,49 +114,120 @@ func (s *DiscoveryServer) StreamListeners(stream xdsapi.ListenerDiscoveryService
+			if err != nil {
+				return err
+			}
 			if !initialRequest {
-				log.Debugf("LDS ACK from Envoy for client %q has version %q and Nonce %q for request", discReq.GetVersionInfo(), discReq.GetResponseNonce())
+				log.Debugf("LDS: ACK from Envoy for client %q has version %q and Nonce %q for request", discReq.GetVersionInfo(), discReq.GetResponseNonce())
 				continue
 			}
+			node.ID = discReq.Node.Id
+			node.Type = nt.Type
+
+			nodeId = nt.ID
+			addLdsCon(nodeId, con)
+			log.Infof("LDS: StreamListeners %v %s %s", peerAddr, nt.ID, discReq.String())
+
 		case <-ticker.C:
 			if !initialRequest {
 				// Ignore ticker events until the very first request is processed.
 				continue
 			}
+		case <-con.pushChannel:
 		}
+
 		if initialRequest {
 			initialRequest = false
-			log.Debugf("LDS request from  %q received.", peerAddr)
 		}
 
-		nt, err := model.ParseServiceNode(discReq.Node.Id)
-		if err != nil {
-			return err
-		}
-		node := model.Proxy{
-			ID:   discReq.Node.Id,
-			Type: nt.Type,
-		}
-		response, err := ListListenersResponse(s.env, node)
-		log.Info(response.String())
+		response, err := con.ldsDiscoveryResponse(s.env, node)
 		if err != nil {
 			return err
 		}
 		err = stream.Send(response)
 		if err != nil {
 			return err
 		}
-		log.Debugf("\nLDS response from  %q, Response: \n%s\n\n", peerAddr, response.String())
 	}
 }
 
-// FetchListeners implements the DiscoveryServer interface.
-func (s *DiscoveryServer) FetchListeners(ctx context.Context, in *xdsapi.DiscoveryRequest) (*xdsapi.DiscoveryResponse, error) {
-	log.Info("FetchListeners")
-	node, err := model.ParseServiceNode(in.Node.Id)
+// ldsPushAll implements old style invalidation, generated when any rule or endpoint changes.
+// Primary code path is from v1 discoveryService.clearCache(), which is added as a handler
+// to the model ConfigStorageCache and Controller.
+func ldsPushAll() {
+	if ldsDebug {
+		log.Infoa("LDS cache reset")
+	}
+	ldsClientsMutex.Lock()
+	// Create a temp map to avoid locking the add/remove
+	tmpMap := map[string]*LdsConnection{}
+	for k, v := range ldsClients {
+		tmpMap[k] = v
+	}
+	version++
+	ldsClientsMutex.Unlock()
+
+	for _, client := range tmpMap {
+		client.pushChannel <- true
+	}
+}
+
+// LDSz implements a status and debug interface for LDS.
+// It is mapped to /debug/ldsz on the monitor port (9093).
+func LDSz(w http.ResponseWriter, req *http.Request) {
+	if req.Form.Get("debug") != "" {
+		ldsDebug = req.Form.Get("debug") == "1"
+		return
+	}
+	if req.Form.Get("push") != "" {
+		ldsPushAll()
+	}
+	ldsClientsMutex.Lock()
+	data, err := json.Marshal(ldsClients)
+	ldsClientsMutex.Unlock()
 	if err != nil {
-		return nil, err
+		_, _ = w.Write([]byte(err.Error()))
+		return
+	}
+
+	_, _ = w.Write(data)
+}
+
+func addLdsCon(s string, connection *LdsConnection) {
+	ldsClientsMutex.Lock()
+	defer ldsClientsMutex.Unlock()
+	ldsClients[s] = connection
+}
+
+func removeLdsCon(connection *LdsConnection, s string) {
```

Comment:
what is the purpose of passing in connection? it seems far fetched that ldsClients[s] != connection so i suggest removing it from the api and remove the check in the function.
otherwise we need an error in the function because clearly something bad has happened.